:fire: remove everything except the cropper and layout analyzer and convert that to OCR-D/core v3 API

kba opened the pull request:
Since this project is largely unmaintained and the only functionality currently used sensibly is the cropper and perhaps the layout analysis, this PR removes everything except those two.

kba committed:
convert cropper to v3 API, setup.py -> pyproject.toml

kba committed:
remove everything that is not cropping-related

kba committed:
adapt cropper and test

kba committed:
CI: remove references to models

kba committed:
fix setuptools packages

kba committed:
remove pix2pixhd

kba committed:
remove old mets files

kba committed:
convert to src-layout

kba committed:
remove constants

bertsky commented on .circleci/config.yml:
```diff
@@ -22,17 +22,6 @@ jobs:
-      - save_cache:
-          paths:
-            - ~/.local/share/ocrd-resources
-          key: v2-models-{{ checksum "ocrd_anybaseocr/ocrd-tool.json" }}
```
Already forgot about that pattern. Too bad it needs to go here.

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_cropping.py:
```diff
@@ -99,11 +87,143 @@ def pil2array(im,alpha=0):
+        # Border now set, image_from_page will return updated image and coords?)
+        cropped_image, cropped_page_xywh, _ = self.workspace.image_from_page(
+            page, page_id, fill='background', transparency=True)
+        alt_image = AlternativeImageType(comments=cropped_page_xywh['features'])
```
```suggestion
        cropped_image, cropped_coords, _ = self.workspace.image_from_page(
            page, page_id, fill='background', transparency=True)
        alt_image = AlternativeImageType(comments=cropped_coords['features'])
```

bertsky commented on src/ocrd_anybaseocr/ocrd-tool.json:
```diff
@@ -0,0 +1,56 @@
+{
+  "version": "1.10.0",
```
should increase

kba replied:
Will do, once merged.

bertsky commented on tests/test_crop.py:
```diff
@@ -1,37 +1,29 @@
+from .assets import assets, copy_of_directory
+
+
+def test_crop():
```
We should also adopt the scheme employed in other v3 processor tests here: checking with METS Server and page-parallel, too. (Infrastructure in conftest.py, test function just gets `processor_kwargs`, including the workspace.)

kba replied:
9d077e2

But my fairly trivial test fails for `pageparallel` and following for `test_layout`, not sure where my logic is wrong :/

bertsky reviewed (commented):
Looks sane.

But I thought we wanted to do that in two separate stages:
1. migrate the cropper to v3, keep the others as-is
2. drop the others, perhaps by archiving the entire repo and re-introducing the cropper elsewhere

At least for the layout-analysis (i.e. model-driven page classifier) here we had the only existing processor of that kind (and to some extent of a tool that writes logical structMap/div, although docstruct now also does). I am not entirely sure the classifier was outright unusable – perhaps we should keep at least that until we have a replacement?

kba committed:
reintroduce layout-analysis

kba committed:
add model to package data

kba committed:
Update src/ocrd_anybaseocr/cli/ocrd_anybaseocr_cropping.py

Co-authored-by: Robert Sachunsky <38561704+bertsky@users.noreply.github.com>

kba committed:
pickle was ignored

kba committed:
Merge branch 'v3-api' of https://github.com/OCR-D/ocrd_anybaseocr into v3-api

kba committed:
CI: larger resource_class

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_cropping.py:
```diff
@@ -62,29 +63,6 @@ def dshow(title):
             os.close(fd)
             plt.clf()
 
-# originally from ocrolib (here also with alpha support):
-def pil2array(im,alpha=0):
-    if im.mode=="L":
-        a = np.frombuffer(im.tobytes(),'B')
-        a.shape = im.height, im.width
-        return a
-    if im.mode=="LA":
-        a = np.frombuffer(im.tobytes(),'B')
-        a.shape = im.height, im.width, 2
-        if not alpha: a = a[:,:,0]
-        return a
-    if im.mode=="RGB":
-        a = np.frombuffer(im.tobytes(),'B')
-        a.shape = im.height, im.width, 3
-        return a
-    if im.mode=="RGBA":
-        a = np.frombuffer(im.tobytes(),'B')
-        a.shape = im.height, im.width, 4
-        if not alpha: a = a[:,:,:3]
-        return a
-    # fallback to Pillow grayscale conversion:
-    return pil2array(im.convert("L"))
-
```
isn't that still used in L164 (now L142)?

kba replied:
Yes, it is now in `src/ocrd_anybaseocr/utils.py`, since both cropper and layout analysis need it.

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_layout_analysis.py:
```diff
@@ -0,0 +1,227 @@
+TAG_METS_SMLINK = '{%s}smLink' % NS['mets']
+
+class OcrdAnybaseocrLayoutAnalyser(Processor):
+    
```
```suggestion
    max_workers = 1 # Tensorflow context cannot be shared across forked workers
```

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_layout_analysis.py:
```diff
@@ -0,0 +1,227 @@
+        
+        assert self.parameter
+        if 'model_path' not in self.parameter:
+            self.parameter = {'model_path': str(resource_filename(f'ocrd_anybaseocr.models', 'structure_analysis')), **self.parameter}
```
I'm afraid you cannot do that anymore. We now have a setter that automatically delegates to `setup` itself.

So for the `setup` function, if you need to mutate the passed dict, just use a copy.

bertsky replied:
Also, why all the hassle in the first place? You could just declare these distributed models as default, and have `self.resolve_resource` magically do the rest below.

kba replied:
Was copying the old pattern from ocrd_froc (which needs update as well), will clean that up.

kba replied:
40f6da8

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_layout_analysis.py:
```diff
@@ -0,0 +1,227 @@
+        class_mapper_path = Path(self.resolve_resource(self.parameter['class_mapping_path']))
+        if not model_path.exists():
+            self.logger.critical("Layout classfication `model_path` was not found at '%s'", model_path)
+            sys.exit(1)
```
This should be removed: `resolve_resource` already raises a `ResourceNotFoundError` if necessary
```suggestion
```

kba replied:
06868dc

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_layout_analysis.py:
```diff
@@ -0,0 +1,227 @@
+        self.model = self.create_model(str(model_path))
+        # load the mapping
+        pickle_in = open(str(class_mapper_path), "rb")
+        class_indices = pickle.load(pickle_in)
```
```suggestion
        with open(str(class_mapper_path), "rb") as pickle_in:
            class_indices = pickle.load(pickle_in)
```

kba replied:
408a400

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_layout_analysis.py:
```diff
@@ -0,0 +1,227 @@
+
+    def process_page_file(self, *input_files : Optional[OcrdFileType]) -> None:
+        self.logger.info("Overridden process_page_file")
+        for input_file in input_files:
```
Instead of this iteration pattern, we should assert our way out of multi-fileGrp input here.

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_layout_analysis.py:
```diff
@@ -0,0 +1,227 @@
+        page = pcgts.get_Page()
+        self.logger.info("INPUT FILE %s", page_id)
+        page_image, _, _ = self.workspace.image_from_page(page, page_id, feature_selector='binarized')
+        img_array = pil2array(page_image.resize((500, 600), Image.LANCZOS))
```
deprecated in PIL
```suggestion
        img_array = pil2array(page_image.resize((500, 600), Image.Resampling.LANCZOS))
```

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_layout_analysis.py:
```diff
@@ -0,0 +1,227 @@
+        self.logger.info(results)
+        #self.workspace.mets.set_physical_page_for_file(input_file.pageId, input_file)
+        self.create_logmap_smlink(pcgts)
+        self.write_to_mets(results, page_id)
+        return result
```
This entire approach does not work with the METS Server, which we cannot disallow beforehand, because `max_workers` only applies to the multiprocessing part. So running with METS Server would always fail in that method (raising something about `ClientSideOcrdMets` not provinding access to some attributes).

But we could do better: as examples like ocrd_pagetopdf or ocrd_cis/ocrd_cis/postcorrect/cli.py illustrate, one can do an extra METS serialization step at the end of `process_workspace` to ensure we get access to the METS file directly (and then have the METS Server reload itself, if present)

bertsky replied:
To be more concrete: on each page, you would just record enough information into a processor attribute dict. Then after the page loop (in `process_workspace`, after the super call), you act on that dict: 
- if necessary (ClientSideOcrdMets): 
  - `self.workspace.mets.save_mets()`, 
  - instantiate a new workspace (to get a direct METS)
  - do the METS update
  - save the new workspace METS
  - `self.workspace.mets.reload_mets()`
- otherwise: do the METS update

kba replied:
e3ba1d7

Not sure whether the final `workspace.mets.reload()` or the `shutdown` are necessary.

bertsky replied:
> Not sure whether the final `workspace.mets.reload()` or the `shutdown` are necessary.

Yes, the `reload` is necessary so the (next processor in the) workflow can proceed with the _correct_ (new) version of the METS.

`shutdown` should actually `del self.model` if it exists. The `reset` should happen at the end of `process_workspace`. And `log_id` / `log_links` / `logID` / `logIDs` do not need to be attributes anymore – they could now be passed in to `write_to_mets`. In contrast, `self.page_labels` can be directly used by that function and does not need to be passed in. In fact, that function could itself loop over it.

kba replied:
e0e63db I moved the `self.reset()` call and changed `self.shutdown()` accordingly, but I don't want to mess with the attributes/method params right now. Feel free to push if you have the stomach for that refactoring :)

bertsky replied:
ok, I'll test locally and see what I can do

bertsky replied:
https://github.com/OCR-D/ocrd_anybaseocr/pull/113

kba committed:
Apply suggestions from code review

Co-authored-by: Robert Sachunsky <38561704+bertsky@users.noreply.github.com>

bertsky committed:
remove redundant model-exists check

kba committed:
Merge branch 'v3-api' of https://github.com/OCR-D/ocrd_anybaseocr into v3-api

bertsky committed:
use contextmanager to open pickled label mapping

kba committed:
rely on resolve_resource for model path

bertsky commented on src/ocrd_anybaseocr/ocrd-tool.json:
```diff
@@ -0,0 +1,136 @@
+      "description": "Generates a table-of-content like document structure of the whole document.",
+      "input_file_grp_cardinality": 1,
+      "output_file_grp_cardinality": 0,
+      "resource_locations": ["module"],
```
Why this restriction?

IIUC just because we currently have no other models does not mean we could not (in theory) add them, as a user.

```suggestion
```

bertsky commented on src/ocrd_anybaseocr/ocrd-tool.json:
```diff
@@ -0,0 +1,136 @@
+          "format": "uri",
+          "content-type": "text/directory",
+          "cacheable": true,
+          "default": "structure_analysis",
```
does not work in the way these are currently packaged (see CI failure)

```suggestion
          "default": "models/structure_analysis",
```

bertsky replied:
Wait. This would work for resolve, but then resmgr list-installed does not show them anymore (because we just disallowed recursion for `module` location in the spec).

What we could instead do in the processor class itself:
```python
    def moduledir(self):
        return resource_filename(self.module, 'models')
```

kba replied:
Or just move the models to the package root. I'll try that first, if it works, I'll rebase before merging, so the git repo stays reasonably small.

bertsky replied:
Of course, that would work as well. But redefining moduledir is the pattern we should generally use, I believe. (Also for preinstalled models in ocrd_froc etc.)

bertsky commented on src/ocrd_anybaseocr/ocrd-tool.json:
```diff
@@ -0,0 +1,136 @@
+          "format": "uri",
+          "content-type": "application/python-pickle",
+          "cacheable": true,
+          "default": "mapping_densenet.pickle",
```
```suggestion
          "default": "models/mapping_densenet.pickle",
```

bertsky reviewed (approved):
Thanks!

(I still think we should make the METS Server scenario workable, though.)

kba committed:
rewrite with process_workspace and deferred METS modification

kba committed:
adapt tests

kba committed:
move models to root of package

kba committed:
self.reset at the end of process_workspace, del self.model in shutdown

bertsky commented on src/ocrd_anybaseocr/cli/ocrd_anybaseocr_layout_analysis.py:
```diff
@@ -96,7 +97,8 @@ def process_page_file(self, *input_files : Optional[OcrdFileType]) -> None:
 
     def shutdown(self) -> None:
-        self.reset()
+        del self.model
+        del self.label_mapping
```
These might never have been initialized, or the GC might already have struck:
```suggestion
        if hasattr(self, 'model'):
            del self.model
        if hasattr(self, 'label_mapping'):
            del self.label_mapping
```

kba committed:
3.8 typing

kba committed:
Be defensive about uninitializing possibly uninitialized model and label_mapping

Co-authored-by: Robert Sachunsky <38561704+bertsky@users.noreply.github.com>

kba committed:
ocrd-tool: don't restrict resource_locations

Co-authored-by: Robert Sachunsky <38561704+bertsky@users.noreply.github.com>

Commit:
test_layout: forgot save (client or direct) + reload (direct)

Commit:
layout-analysis: forgot to save_mets

Commit:
layout-analysis: whitespace + readability + some logging

Commit:
layout-analysis: dedicated moduledir 'models'

Commit:
layout-analysis: rename+refactor div creation

Commit:
layout-analysis: improve log experience

Commit:
layout-analysis: if logical struct exists, add to its deepest top-level div

Commit:
docker: update+improve…

- rebase on ocrd v3.3 image
- add labels and variables according to spec
- preinstall ocrd-all-tool.json
- simplify

Commit:
makefile: improve

Commit:
update CI

Commit:
layout-analysis: reformatted model to suit tf.keras>=2.5

Commit:
forgot to add build rule

Commit:
layout-analysis: do not pre-allocate GPUs

kba committed:
Merge pull request #113 from bertsky/v3-api

finish v3 api
